go/vt/wrangler/reparent.go

```diff
@@ -996,55 +995,279 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
+		filePositionList := make([]mysql.Position, 0, len(replicationStatusMap))
+		for _, p := range positionMap {
+			filePositionList = append(filePositionList, p)
+		}
+		if !mysql.AllPositionsComparable(filePositionList) {
+			return nil, fmt.Errorf("we can't compare all file based positions")
+		}
+	}
+
+	// Find a position that is fully caught up when errant GTIDs are taken out of the equation.
+	var winningPosition mysql.Position
+	for _, position := range positionMap {
+		if winningPosition.IsZero() {
+			winningPosition = position
+		}
+		if position.AtLeast(winningPosition) {
+			winningPosition = position
+		}
+	}
+
+	// Now that we have a high water mark, we can get a list of all tablets that are errant GTID free
+	// that hit this high water mark.
+	var validCandidates []string
+	for alias, position := range positionMap {
+		if badTablets.Has(alias) {
+			// We should exclude tablets we've already determined have errant GTIDs from consideration.
+			continue
+		}
+
+		if position.AtLeast(winningPosition) {
+			validCandidates = append(validCandidates, alias)
+		}
+	}
+
+	return validCandidates, nil
+}
+
+func (wr *Wrangler) stopReplicationAndBuildStatusMaps(ctx context.Context, ev *events.Reparent, tabletMap map[string]*topo.TabletInfo, waitReplicasTimeout time.Duration, unreachableReplicas sets.String) (map[string]*replicationdatapb.StopReplicationStatus, map[string]*replicationdatapb.MasterStatus, error) {
+	// Stop replication on all replicas, get their current
+	// replication position
+	event.DispatchUpdate(ev, "stop replication on all replicas")
+	statusMap := make(map[string]*replicationdatapb.StopReplicationStatus)
+	masterStatusMap := make(map[string]*replicationdatapb.MasterStatus)
+	mu := sync.Mutex{}
+
+	errChan := make(chan error)
+	groupCtx, groupCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+	defer groupCancel()
+	fillStatus := func(alias string, tabletInfo *topo.TabletInfo) {
+		err := fmt.Errorf("fillStatus did not successfully complete")
+		returnErr := func() { errChan <- err }
+		defer returnErr()
+
+		wr.logger.Infof("getting replication position from %v", alias)
+		ctx, cancel := context.WithCancel(groupCtx)
+		defer cancel()
+		var stopReplicationStatus *replicationdatapb.StopReplicationStatus
+		_, stopReplicationStatus, err = wr.tmc.StopReplicationAndGetStatus(ctx, tabletInfo.Tablet, replicationdatapb.StopReplicationMode_IOTHREADONLY)
+		switch err {
+		case mysql.ErrNotReplica:
+			fmt.Printf("Found ErrNotReplica for alias: %v", alias)
+			masterStatus, err := wr.tmc.DemoteMaster(ctx, tabletInfo.Tablet)
+			if err != nil {
+				wr.logger.Warningf("replica %v thinks it's master but we failed to demote it", alias)
+				return
+			}
+			mu.Lock()
+			masterStatusMap[alias] = masterStatus
+			mu.Unlock()
+
+		case nil:
+			mu.Lock()
+			statusMap[alias] = stopReplicationStatus
+			mu.Unlock()
+
+		default:
+			wr.logger.Warningf("failed to get replication status from %v, ignoring tablet: %v", alias, err)
```

Comment:
It's a bit misleading to say here that we are ignoring the tablet. If more than one tablet hits this code path, we won't ignore them.

Reply:
That's true, I'll refactor the message.

Reply:
Done.